gobblin-core-base/src/main/java/gobblin/writer/WatermarkManager.java

```diff
@@ -0,0 +1,108 @@
+
+import gobblin.source.extractor.ComparableWatermark;
+import gobblin.util.ExecutorsUtils;
+
+
+/**
+ * Responsible for managing continuous commit of watermarks.
+ * Periodically fetches watermarks from WatermarkAwareWriters and commits them to WatermarkStorage.
+ */
+public class WatermarkManager implements Closeable {
+
+  private final List<WatermarkAwareWriter> _watermarkAwareWriters;
+  private final WatermarkStorage _watermarkStorage;
+  private final long _commitIntervalMillis;
+  private final ScheduledExecutorService _watermarkCommitThreadPool;
+  private final Logger _logger;
+
+  @VisibleForTesting
+  final Runnable _watermarkCommitter = new
+
+      Runnable() {
+        @Override
+        public void run() {
+
+          ComparableWatermark watermarkToCommit = null;
+          for (WatermarkAwareWriter writer : _watermarkAwareWriters) {
+            Optional<ComparableWatermark> optionalWriterWatermark = writer.getCommittableWatermark();
+            if (optionalWriterWatermark.isPresent()) {
+              ComparableWatermark writerWatermark = optionalWriterWatermark.get();
+              _logger.debug("Found watermark {} from writer {}", writerWatermark.toString(), writer.getClass());
+              if (watermarkToCommit == null) {
+                watermarkToCommit = writerWatermark;
+              } else {
+                watermarkToCommit =
+                    (watermarkToCommit.compareTo(writerWatermark) < 0) ? watermarkToCommit : writerWatermark;
+              }
+            } else {
+              _logger.info("Found absent watermark from writer {}, nothing to commit", writer.getClass());
+              return;
+            }
+          }
+
+          if (watermarkToCommit != null) {
+            try {
+              _logger.info("Will commit watermark {}", watermarkToCommit.toString());
+              _watermarkStorage.commitWatermark(watermarkToCommit);
+            } catch (Exception e) {
+              _logger.error("Failed to write watermark", e);
+            }
+          }
+        }
+      };
+
+
+  public WatermarkManager(WatermarkStorage storage, long commitIntervalMillis, Optional<Logger> logger) {
+    _watermarkAwareWriters = new ArrayList<>(1);
+    _watermarkStorage = storage;
+    _commitIntervalMillis = commitIntervalMillis;
+    _logger = logger.or(LoggerFactory.getLogger(WatermarkManager.class));
+    _watermarkCommitThreadPool = new ScheduledThreadPoolExecutor(1, ExecutorsUtils.newThreadFactory(logger));
+  }
+
+  //TODO: Figure out contract, should register writer only be called with WatermarkAwareWriter?
+  public void registerWriter(WatermarkAwareWriter dataWriter) {
+      _watermarkAwareWriters.add(dataWriter);
+      _logger.info("Registered a watermark aware writer {}");
+  }
+
+  public void start() {
+    _watermarkCommitThreadPool
+        .scheduleWithFixedDelay(_watermarkCommitter, 0, _commitIntervalMillis, TimeUnit.MILLISECONDS);
+  }
+
+  @Override
+  public void close()
```

Comment:
doesn't seem to be invoked anywhere, I was expecting this to be invoked via some sort of finishJob() / finishTask() and the last write of writers consolidated to update final watermark